File: android/app/src/main/java/com/nurio/android/payments/PaymentCrashContext.kt
package com.nurio.android.payments

import com.google.firebase.crashlytics.FirebaseCrashlytics
import java.security.MessageDigest

internal interface PaymentCrashReporter {
    fun setBoolean(key: String, value: Boolean)
    fun setString(key: String, value: String)
    fun log(message: String)
    fun record(exception: Throwable)
}

private class FirebasePaymentCrashReporter(
    private val crashlytics: FirebaseCrashlytics = FirebaseCrashlytics.getInstance(),
) : PaymentCrashReporter {
    override fun setBoolean(key: String, value: Boolean) = crashlytics.setCustomKey(key, value)
    override fun setString(key: String, value: String) = crashlytics.setCustomKey(key, value)
    override fun log(message: String) = crashlytics.log(message)
    override fun record(exception: Throwable) = crashlytics.recordException(exception)
}

internal enum class PaymentStage(val wireValue: String) {
    CHECKOUT_PRESENTED("checkout_presented"),
    SDK_READY("sdk_ready"),
    PAYMENT_REQUESTED("payment_requested"),
    ATTEMPT_REFRESHED("attempt_refreshed"),
    GATEWAY_HANDOFF("gateway_handoff"),
    EXTERNAL_APP_HANDOFF("external_app_handoff"),
    CALLBACK_RECEIVED("callback_received"),
    COMPLETION_VERIFYING("completion_verifying"),
    RETRY_COLD_BOOT("retry_cold_boot"),
    FLOW_FINISHED("flow_finished");

    companion object {
        fun fromWireValue(value: String?): PaymentStage? = entries.firstOrNull { it.wireValue == value }
    }
}

internal enum class PaymentFailureKind(val wireValue: String, val code: Int) {
    NONE("none", 0),
    CONFIGURATION("configuration", 1),
    SDK_LOAD("sdk_load", 2),
    SDK_REQUEST("sdk_request", 3),
    ATTEMPT_REFRESH("attempt_refresh", 4),
    NATIVE_REQUEST("native_request", 5),
    MALFORMED_INTENT("malformed_intent", 6),
    EXTERNAL_APP_UNAVAILABLE("external_app_unavailable", 7),
    EXTERNAL_APP_SECURITY("external_app_security", 8),
    MALFORMED_CALLBACK("malformed_callback", 9),
    RENDERER_TERMINATED("renderer_terminated", 10);

    companion object {
        fun fromWireValue(value: String?): PaymentFailureKind =
            entries.firstOrNull { it.wireValue == value } ?: NONE
    }
}

internal class PaymentTechnicalException(failure: PaymentFailureKind) :
    RuntimeException("payment_failure:${failure.wireValue}")

internal class PaymentCrashContext(
    private val reporter: PaymentCrashReporter,
    private val appSurface: String,
) {
    private var active = false
    private var stage = "none"
    private var orderKind = "unknown"
    private var reference = "none"
    private var handoff = "none"
    private var failureKind = PaymentFailureKind.NONE

    fun reset() {
        active = false
        stage = "none"
        orderKind = "unknown"
        reference = "none"
        handoff = "none"
        failureKind = PaymentFailureKind.NONE
        applyKeys()
    }

    fun track(
        stageValue: String?,
        orderKindValue: String?,
        paymentReference: String?,
        handoffValue: String?,
        failureKindValue: String?,
        reportNonfatal: Boolean,
    ) {
        val nextStage = PaymentStage.fromWireValue(stageValue) ?: return
        if (nextStage == PaymentStage.FLOW_FINISHED) {
            safeLog("payment:flow_finished")
            reset()
            return
        }

        active = true
        stage = nextStage.wireValue
        orderKind = sanitizeOrderKind(orderKindValue)
        paymentReference?.takeIf(String::isNotBlank)?.let { reference = hashReference(it) }
        handoff = sanitizeHandoff(handoffValue)
        failureKind = PaymentFailureKind.fromWireValue(failureKindValue)
        applyKeys()
        safeLog("payment:${nextStage.wireValue}")

        if (reportNonfatal && failureKind != PaymentFailureKind.NONE) {
            safeRecord(failureKind)
        }
    }

    fun markExternalAppHandoff() {
        track(
            stageValue = PaymentStage.EXTERNAL_APP_HANDOFF.wireValue,
            orderKindValue = orderKind,
            paymentReference = null,
            handoffValue = "external_app",
            failureKindValue = PaymentFailureKind.NONE.wireValue,
            reportNonfatal = false,
        )
    }

    fun markCallbackReceived(paymentReference: String?) {
        track(
            stageValue = PaymentStage.CALLBACK_RECEIVED.wireValue,
            orderKindValue = orderKind,
            paymentReference = paymentReference,
            handoffValue = "native_callback",
            failureKindValue = PaymentFailureKind.NONE.wireValue,
            reportNonfatal = false,
        )
    }

    fun logRetryColdBoot() {
        if (active) safeLog("payment:${PaymentStage.RETRY_COLD_BOOT.wireValue}")
    }

    fun reportTechnicalFailure(failure: PaymentFailureKind) {
        active = true
        failureKind = failure
        applyKeys()
        safeLog("payment_failure:${failure.wireValue}")
        safeRecord(failure)
    }

    fun reportRendererTermination() {
        if (active) reportTechnicalFailure(PaymentFailureKind.RENDERER_TERMINATED)
    }

    fun isActive(): Boolean = active

    internal fun hashReference(value: String): String {
        return try {
            MessageDigest.getInstance("SHA-256")
                .digest(value.toByteArray(Charsets.UTF_8))
                .joinToString(separator = "") { byte -> "%02x".format(byte) }
                .take(16)
        } catch (_: Exception) {
            "unavailable"
        }
    }

    private fun applyKeys() {
        safeSetString("app_surface", appSurface)
        safeSetString("native_platform", "android")
        safeSetBoolean("payment_flow_active", active)
        safeSetString("payment_stage", stage)
        safeSetString("payment_order_kind", orderKind)
        safeSetString("payment_reference", reference)
        safeSetString("payment_provider", if (active) "portone_inicis" else "none")
        safeSetString("payment_handoff", handoff)
        safeSetString("payment_failure_kind", failureKind.wireValue)
    }

    private fun safeSetBoolean(key: String, value: Boolean) = bestEffort {
        reporter.setBoolean(key, value)
    }

    private fun safeSetString(key: String, value: String) = bestEffort {
        reporter.setString(key, value)
    }

    private fun safeLog(message: String) = bestEffort { reporter.log(message) }

    private fun safeRecord(failure: PaymentFailureKind) = bestEffort {
        reporter.record(PaymentTechnicalException(failure))
    }

    private inline fun bestEffort(action: () -> Unit) {
        try {
            action()
        } catch (_: Exception) {
            // Crash reporting is optional and must never affect payment behavior.
        }
    }

    private fun sanitizeOrderKind(value: String?): String = when (value) {
        "ticket", "pass_package", "deposit", "study_group", "event_add_on" -> value
        else -> "unknown"
    }

    private fun sanitizeHandoff(value: String?): String = when (value) {
        "webview", "external_app", "native_callback", "none" -> value
        else -> "none"
    }
}

internal object PaymentCrashTelemetry {
    private val context by lazy {
        PaymentCrashContext(
            reporter = FirebasePaymentCrashReporter(),
            appSurface = "nurio",
        )
    }

    fun reset() = bestEffort { context.reset() }

    fun track(
        stage: String?,
        orderKind: String?,
        paymentReference: String?,
        handoff: String?,
        failureKind: String?,
        reportNonfatal: Boolean,
    ) = bestEffort {
        context.track(stage, orderKind, paymentReference, handoff, failureKind, reportNonfatal)
    }

    fun markExternalAppHandoff() = bestEffort { context.markExternalAppHandoff() }
    fun markCallbackReceived(paymentReference: String?) = bestEffort {
        context.markCallbackReceived(paymentReference)
    }
    fun logRetryColdBoot() = bestEffort { context.logRetryColdBoot() }
    fun reportTechnicalFailure(failure: PaymentFailureKind) = bestEffort {
        context.reportTechnicalFailure(failure)
    }
    fun reportRendererTermination() = bestEffort { context.reportRendererTermination() }
    fun isActive(): Boolean = bestEffort(default = false) { context.isActive() }

    private inline fun bestEffort(action: () -> Unit) {
        try {
            action()
        } catch (_: Exception) {
            // Firebase initialization/reporting must never affect the app flow.
        }
    }

    private inline fun <T> bestEffort(default: T, action: () -> T): T {
        return try {
            action()
        } catch (_: Exception) {
            default
        }
    }
}

File: ios/Payments/PaymentCrashContext.swift
import CryptoKit
import FirebaseCrashlytics
import Foundation

protocol PaymentCrashReporting {
    func setCustomValue(_ value: Any, forKey key: String) throws
    func log(_ message: String) throws
    func record(error: Error) throws
}

private final class FirebasePaymentCrashReporter: PaymentCrashReporting {
    private let crashlytics: Crashlytics

    init(crashlytics: Crashlytics = .crashlytics()) {
        self.crashlytics = crashlytics
    }

    func setCustomValue(_ value: Any, forKey key: String) throws {
        crashlytics.setCustomValue(value, forKey: key)
    }

    func log(_ message: String) throws {
        crashlytics.log(message)
    }

    func record(error: Error) throws {
        crashlytics.record(error: error)
    }
}

enum PaymentFailureKind: String {
    case none
    case configuration
    case sdkLoad = "sdk_load"
    case sdkRequest = "sdk_request"
    case attemptRefresh = "attempt_refresh"
    case nativeRequest = "native_request"
    case malformedIntent = "malformed_intent"
    case externalAppUnavailable = "external_app_unavailable"
    case externalAppSecurity = "external_app_security"
    case malformedCallback = "malformed_callback"
    case rendererTerminated = "renderer_terminated"

    var code: Int {
        switch self {
        case .none: 0
        case .configuration: 1
        case .sdkLoad: 2
        case .sdkRequest: 3
        case .attemptRefresh: 4
        case .nativeRequest: 5
        case .malformedIntent: 6
        case .externalAppUnavailable: 7
        case .externalAppSecurity: 8
        case .malformedCallback: 9
        case .rendererTerminated: 10
        }
    }
}

final class PaymentCrashContext {
    private let reporter: PaymentCrashReporting
    private let appSurface: String

    private(set) var isActive = false
    private var stage = "none"
    private var orderKind = "unknown"
    private var reference = "none"
    private var handoff = "none"
    private var failureKind = PaymentFailureKind.none

    init(reporter: PaymentCrashReporting, appSurface: String) {
        self.reporter = reporter
        self.appSurface = appSurface
    }

    func reset() {
        isActive = false
        stage = "none"
        orderKind = "unknown"
        reference = "none"
        handoff = "none"
        failureKind = .none
        applyKeys()
    }

    func track(
        stage stageValue: String?,
        orderKind orderKindValue: String?,
        paymentReference: String?,
        handoff handoffValue: String?,
        failureKind failureKindValue: String?,
        reportNonfatal: Bool
    ) {
        guard Self.allowedStages.contains(stageValue ?? "") else { return }
        guard let stageValue else { return }

        if stageValue == "flow_finished" {
            safeLog("payment:flow_finished")
            reset()
            return
        }

        isActive = true
        stage = stageValue
        if let orderKindValue, Self.allowedOrderKinds.contains(orderKindValue) {
            orderKind = orderKindValue
        } else {
            orderKind = "unknown"
        }
        if let paymentReference, !paymentReference.isEmpty {
            reference = hashReference(paymentReference)
        }
        if let handoffValue, Self.allowedHandoffs.contains(handoffValue) {
            handoff = handoffValue
        } else {
            handoff = "none"
        }
        failureKind = PaymentFailureKind(rawValue: failureKindValue ?? "") ?? .none
        applyKeys()
        safeLog("payment:\(stageValue)")

        if reportNonfatal, failureKind != .none {
            safeRecord(failureKind)
        }
    }

    func markExternalAppHandoff() {
        track(
            stage: "external_app_handoff",
            orderKind: orderKind,
            paymentReference: nil,
            handoff: "external_app",
            failureKind: PaymentFailureKind.none.rawValue,
            reportNonfatal: false
        )
    }

    func markCallbackReceived(paymentReference: String?) {
        track(
            stage: "callback_received",
            orderKind: orderKind,
            paymentReference: paymentReference,
            handoff: "native_callback",
            failureKind: PaymentFailureKind.none.rawValue,
            reportNonfatal: false
        )
    }

    func logRetryColdBoot() {
        guard isActive else { return }
        safeLog("payment:retry_cold_boot")
    }

    func reportTechnicalFailure(_ failure: PaymentFailureKind) {
        isActive = true
        failureKind = failure
        applyKeys()
        safeLog("payment_failure:\(failure.rawValue)")
        safeRecord(failure)
    }

    func reportNativeRequestFailure() {
        guard isActive else { return }
        reportTechnicalFailure(.nativeRequest)
    }

    func hashReference(_ value: String) -> String {
        let digest = SHA256.hash(data: Data(value.utf8))
        return digest.prefix(8).map { String(format: "%02x", $0) }.joined()
    }

    private func applyKeys() {
        safeSet(appSurface, forKey: "app_surface")
        safeSet("ios", forKey: "native_platform")
        safeSet(isActive, forKey: "payment_flow_active")
        safeSet(stage, forKey: "payment_stage")
        safeSet(orderKind, forKey: "payment_order_kind")
        safeSet(reference, forKey: "payment_reference")
        safeSet(isActive ? "portone_inicis" : "none", forKey: "payment_provider")
        safeSet(handoff, forKey: "payment_handoff")
        safeSet(failureKind.rawValue, forKey: "payment_failure_kind")
    }

    private func safeSet(_ value: Any, forKey key: String) {
        bestEffort { try reporter.setCustomValue(value, forKey: key) }
    }

    private func safeLog(_ message: String) {
        bestEffort { try reporter.log(message) }
    }

    private func safeRecord(_ failure: PaymentFailureKind) {
        let error = NSError(
            domain: "com.nurio.payment",
            code: failure.code,
            userInfo: [NSLocalizedDescriptionKey: "payment_failure:\(failure.rawValue)"]
        )
        bestEffort { try reporter.record(error: error) }
    }

    private func bestEffort(_ action: () throws -> Void) {
        do {
            try action()
        } catch {
            // Crash reporting is optional and must never affect payment behavior.
        }
    }

    private static let allowedStages: Set<String> = [
        "checkout_presented",
        "sdk_ready",
        "payment_requested",
        "attempt_refreshed",
        "gateway_handoff",
        "external_app_handoff",
        "callback_received",
        "completion_verifying",
        "retry_cold_boot",
        "flow_finished",
    ]

    private static let allowedOrderKinds: Set<String> = [
        "ticket",
        "pass_package",
        "deposit",
        "study_group",
        "event_add_on",
        "unknown",
    ]

    private static let allowedHandoffs: Set<String> = [
        "webview",
        "external_app",
        "native_callback",
        "none",
    ]
}

enum PaymentCrashTelemetry {
    private static var context: PaymentCrashContext?

    static func configure(appSurface: String) {
        let configuredContext = PaymentCrashContext(
            reporter: FirebasePaymentCrashReporter(),
            appSurface: appSurface
        )
        context = configuredContext
        configuredContext.reset()
    }

    static func track(
        stage: String?,
        orderKind: String?,
        paymentReference: String?,
        handoff: String?,
        failureKind: String?,
        reportNonfatal: Bool
    ) {
        context?.track(
            stage: stage,
            orderKind: orderKind,
            paymentReference: paymentReference,
            handoff: handoff,
            failureKind: failureKind,
            reportNonfatal: reportNonfatal
        )
    }

    static func markExternalAppHandoff() {
        context?.markExternalAppHandoff()
    }

    static func markCallbackReceived(paymentReference: String?) {
        context?.markCallbackReceived(paymentReference: paymentReference)
    }

    static func logRetryColdBoot() {
        context?.logRetryColdBoot()
    }

    static func reportTechnicalFailure(_ failure: PaymentFailureKind) {
        context?.reportTechnicalFailure(failure)
    }

    static func reportNativeRequestFailure() {
        context?.reportNativeRequestFailure()
    }
}
